File: code/大臺北公車 資料爬蟲.R
library(dplyr)
library(sf)
library(rvest)
library(rjson)

# 擷取所有路線表
url="https://ebus.gov.taipei/ebus"
html_content=read_html(url)
all_route=html_nodes(html_content, "#list a")
all_route=data.frame(UniRouteId=gsub("javascript:go\\(|\\)|'", "", html_attr(all_route, 'href')),
                     UniRouteName=gsub(" ", "", html_text(all_route)))


# 逐一擷取路線與站點
bus_route_all=data.frame()
bus_stop_all=data.frame()
bus_shape_all=data.frame()

for(i in c(1:length(all_route$UniRouteId))){
  url=paste0("https://ebus.gov.taipei/Route/StopsOfRoute?routeid=", all_route$UniRouteId[i])
  html_content=read_html(url)
  html_content=html_nodes(html_content, "script")
  html_content=html_content[grepl("routeJsonString", html_content)]
  html_content=strsplit(as.character(html_content), "\n")[[1]]
  html_content=html_content[grepl("routeJsonString", html_content)]
  html_content=strsplit(as.character(html_content), ";\\r")[[1]]
  html_content=as.character(html_content[grepl("stringify", html_content)])
  html_content=substr(html_content, regexpr("stringify", html_content)+10, max(which(strsplit(html_content, "")[[1]]=="}")))
  html_content=rjson::fromJSON(html_content)
  
  # Bus Route
  bus_route=data.frame(html_content[c("UniRouteId","Name","BeginStop","LastStop","TicketPrice","BusTimeDesc","HeadwayDesc","HolidayBusTimeDesc","HolidayHeadwayDesc","SegmentBuffer","orgphone")])%>%
    rename(UniRouteName=Name)
  
  # Bus Stop
  temp1=data.frame(UniRouteId=bus_route$UniRouteId,
                   UniRouteName=bus_route$UniRouteName,
                   UniStopId=mapply(function(x) html_content$GoDirStops[[x]]$UniStopId, c(1:length(html_content$GoDirStops))),
                   UniStopName=mapply(function(x) html_content$GoDirStops[[x]]$Name, c(1:length(html_content$GoDirStops))),
                   Longitude=mapply(function(x) html_content$GoDirStops[[x]]$Longitude, c(1:length(html_content$GoDirStops))),
                   Latitude=mapply(function(x) html_content$GoDirStops[[x]]$Latitude, c(1:length(html_content$GoDirStops))),
                   Direction=1,
                   StopSequence=c(1:length(html_content$GoDirStops)))
  if(!is.null(html_content$BackDirStops)){
    temp2=data.frame(UniRouteId=bus_route$UniRouteId,
                     UniRouteName=bus_route$UniRouteName,
                     UniStopId=mapply(function(x) html_content$BackDirStops[[x]]$UniStopId, c(1:length(html_content$BackDirStops))),
                     UniStopName=mapply(function(x) html_content$BackDirStops[[x]]$Name, c(1:length(html_content$BackDirStops))),
                     Longitude=mapply(function(x) html_content$BackDirStops[[x]]$Longitude, c(1:length(html_content$BackDirStops))),
                     Latitude=mapply(function(x) html_content$BackDirStops[[x]]$Latitude, c(1:length(html_content$BackDirStops))),
                     Direction=2,
                     StopSequence=c(1:length(html_content$BackDirStops)))
    bus_stop=rbind(temp1, temp2)
  }else{
    bus_stop=temp1
  }
  
  # Bus Shape
  bus_shape=data.frame(UniRouteId=bus_route$UniRouteId,
                       UniRouteName=bus_route$UniRouteName,
                       WKT=c("wkt","wkt0","wkt1"),
                       geometry=c(ifelse(is.null(html_content$wkt), NA, html_content$wkt),
                                  ifelse(is.null(html_content$wkt0), NA, html_content$wkt0),
                                  ifelse(is.null(html_content$wkt1), NA, html_content$wkt1)))
  
  bus_route_all=rbind(bus_route_all, bus_route)
  bus_stop_all=rbind(bus_stop_all, bus_stop)
  bus_shape_all=rbind(bus_shape_all, bus_shape)
  print(i)
}

File: code/大臺北公車 資料爬蟲.py
import pandas as pd
import numpy as np
import geopandas as gpd
from shapely import wkt
import requests
from bs4 import BeautifulSoup
import json
import re
from tqdm import tqdm


# 擷取所有路線表
res=requests.get('https://ebus.gov.taipei/ebus')
res=BeautifulSoup(res.text, "html.parser")
res=res.find_all("section", class_="busline")
res=BeautifulSoup(str(res))
res=res.find_all('a')

routeid=[]
routename=[]
for route in res:
    routeid.append(re.sub("javascript:go\\(|\\)|'", '', route.get("href")))
    routename.append(re.sub(' ', '', route.get_text()))

all_route=pd.DataFrame({'UniRouteId':routeid, 'UniRouteName':routename})    
all_route=all_route.drop_duplicates().reset_index(drop=True)



# 逐一擷取路線與站點
bus_route_all=pd.DataFrame()
bus_stop_all=pd.DataFrame()
bus_shape_all=pd.DataFrame()

for i in tqdm(range(len(all_route))):
    res=requests.get('https://ebus.gov.taipei/Route/StopsOfRoute?routeid='+all_route.UniRouteId[i])
    res=BeautifulSoup(res.text, "html.parser")
    res=res.find_all("script")
    res=str(list(compress(res, [bool(re.search("routeJsonString", str(x))) for x in res])))
    res=res.split('\n')
    res=str(list(compress(res, [bool(re.search("routeJsonString", x)) for x in res])))
    res=res.split(';\\r')
    res=str(list(compress(res, [bool(re.search("stringify", x)) for x in res])))
    res=res[(re.search("stringify", res).span()[1]+1):(res.rfind('}')+1)]
    res=json.loads(res)

    # Bus Stop
    temp1=pd.DataFrame(res['GoDirStops']).loc[:,['UniStopId','Name','Longitude','Latitude']].rename(columns={'Name':'UniStopName'})
    temp1['Direction']=1
    temp1['StopSequence']=list(range(1, len(temp1)+1))
    
    if res['BackDirStops'] is not None:
        temp2=pd.DataFrame(res['BackDirStops']).loc[:,['UniStopId','Name','Longitude','Latitude']].rename(columns={'Name':'UniStopName'})
        temp2['Direction']=2
        temp2['StopSequence']=list(range(1, len(temp2)+1))
        bus_stop=pd.concat([temp1, temp2]).reset_index(drop=True)
    else:
        bus_stop=temp1
    
    # Bus Route
    bus_route=pd.DataFrame({'UniRouteId':[res['UniRouteId']],
                            'UniRouteName':[res['Name']],
                            'BeginStop':[res['BeginStop']],
                            'LastStop':[res['LastStop']],
                            'TicketPrice':res['TicketPrice'],
                            'BusTimeDesc':res['BusTimeDesc'],
                            'HeadwayDesc':res['HeadwayDesc'],
                            'HolidayBusTimeDesc':res['HolidayBusTimeDesc'],
                            'HolidayHeadwayDesc':res['HolidayHeadwayDesc'],
                            'SegmentBuffer':res['SegmentBuffer'],
                            'orgphone':res['orgphone']})
    bus_stop=pd.concat([bus_route.loc[np.repeat(0, len(bus_stop)), ['UniRouteId','UniRouteName']].reset_index(drop=True), bus_stop], axis=1)
    
    # Bus Shape
    bus_shape=gpd.GeoDataFrame(pd.concat([bus_route.loc[np.repeat(0, 3), ['UniRouteId','UniRouteName']].reset_index(drop=True), pd.DataFrame({'WKT':['wkt','wkt0','wkt1']})], axis=1),
                               geometry=gpd.GeoSeries.from_wkt([res['wkt'], res['wkt0'], res['wkt1']]), crs=4326)
    
    bus_stop_all=pd.concat([bus_stop_all, bus_stop]).reset_index(drop=True)
    bus_shape_all=pd.concat([bus_shape_all, bus_shape]).reset_index(drop=True)
    bus_route_all=pd.concat([bus_route_all, bus_route]).reset_index(drop=True)
    

